scripts/game_logic/grid.c

```diff
@@ -0,0 +1,123 @@
+/*
+ ============================================================================
+ Name        : grid_gen.c
+ Author      : Liam Tapper
+ Version     :
+ Copyright   :
+ Description : General Idea for grid generation and traversal while being O(1)
+         or as close as possible to it.
+         This just holds basic game logic
+ ============================================================================
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include "grid.h"
+/*
+ * TODO : learn how turns work to create a 'move queue'
+ *
+ * TODO : REMOVE STDIO AND STDLIB WHEN FINISHED
+ *
+ * TODO : Create new branch + PR for character movement and turn queues
+ */
+
+
+int main(void) {
+  struct GameManager gm;
+  gm.grid = CreateTiles(gm.grid);
+  gm = InitializeCharacter(2,7,gm);
+  printf("newCharacterIndex: %i\nx:%i\ny:%i\n",
+      gm.characters[0].index, FindX(gm.characters[0].index), FindY(gm.characters[0].index));
+  PrintBoard(gm.grid);
+  return EXIT_SUCCESS;
+}
+
+/*
+ * Generates the tiles and inserts them into the grid struct
+ * Note: (i * COL + j) is a dumb formula that ensures uniqueness
+ * Pretty sure this still qualifies as O(1) due to the static nature
+ * of ROWS and COL (constant iteration count).
+ */
+struct Grid CreateTiles(struct Grid grid){
+  for(int i = 1; i <= ROWS; i++){
+    for (int j = 1; j <= COL; j++){
+      struct Tile tile;
+      tile.index = (i*COL+j)-COL;
+      grid.tiles[tile.index] = tile;
+    }
+  }
+  struct Tile specialTile = { .index = 0 };
+  grid.tiles[0] = specialTile;
+  return grid;
+}
+/*
+ * Finds the x and y cords on the board given the index of the tiles
+ * within the grid array
+ * Say we throw index 17 into Find x and y,
+ * it will spit out the row and collumn for the given grid array index :)
+ */
+int FindX(int index){
+  return (index == 0) ? 0 : ((index-1)/COL+1);
+}
+
+int FindY(int index){
+  return (index == 0) ? 0 : ((index-1)%COL+1);
+}
+
+/*
+ * Character Initialization
+ * A return character of index -1 marks failed initialization
+ */
+struct GameManager InitializeCharacter(int x, int y, struct GameManager gm){
+  struct Character newCharacter = { .index = -1 };
+  if(CheckTile(x,y,gm.grid) == -1){
+    return gm;
+  }
```

Comment:
Normally would want to print error statement. 

For this and other "prints" define a print function and for now use the stdlib, but for embedded it will be different but make the skeleton for it now.

Sort of like this

#define IS_EMBEDDED 0
def print(string) {
  if (IS_EMBEDDED) {
    // write over uart
  }
  else {
    printf(string);
  }
}

do this similarily for importing the stdio and stdlib

Reply:
I would recommend adding that to a new .h file., (print.h idek)